pkg/agent/util/net_linux.go

```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
 	cmd := exec.Command("ovs-vsctl", "--if-exists", "del-port", brName, portName)
 	return cmd.Run()
 }
+
+func HostInterfaceExists(ifName string) bool {
+	_, err := netlink.LinkByName(ifName)
+	if err == nil {
+		return true
+	}
+	if _, ok := err.(netlink.LinkNotFoundError); ok {
+		return false
+	}
+	klog.ErrorS(err, "Failed to find host interface", "name", ifName)
+	return false
+}
+
+func GetUplinkConfig(uplinkIfName string) (*config.AdapterNetConfig, error) {
+	iface, err := net.InterfaceByName(uplinkIfName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get interface", "uplinkIfName", uplinkIfName)
+		return nil, err
+	}
+	addrs, err := GetGlobalIPNetsByName(iface)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get address", "iface", iface)
+		return nil, err
+	}
+	gw, routes, err := getRoutesAndDefaultGwOnInterface(iface.Index)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get routes", "iface.Index", iface.Index)
+		return nil, err
+	}
+	return &config.AdapterNetConfig{
+		Name:    uplinkIfName,
+		Index:   iface.Index,
+		MAC:     iface.HardwareAddr,
+		IPs:     addrs,
+		Routes:  routes,
+		Gateway: gw,
+		MTU:     iface.MTU,
+	}, nil
+}
+
+func RenameInterface(from, to string) error {
+	var renameErr error
+	// Add retry here because it might return the error of device is busy for one time configuration.
+	pollErr := wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
+		renameErr = renameHostInterface(from, to)
+		if renameErr != nil {
+			klog.InfoS("Unable to rename host interface name, retrying", "oldName", from, "newName", to, "err", renameErr)
+			return false, nil
+		}
+		return true, nil
+	})
+	if pollErr != nil {
+		return fmt.Errorf("failed to rename host interface name %s to %s", from, to)
+	}
+	return nil
+}
+
+func MoveIFConfigurations(ips []*net.IPNet, routes []interface{}, mtu int, from string, to string) error {
+	toIF, err := netlink.LinkByName(to)
+	if err != nil {
+		return err
+	}
+	fromExists := true
+	fromIF, err := netlink.LinkByName(from)
+	if err != nil {
+		if _, ok := err.(netlink.LinkNotFoundError); !ok {
+			return err
+		}
+		fromExists = false
+	}
+	if fromExists {
+		if err := netlink.LinkSetMTU(toIF, mtu); err != nil {
+			return err
+		}
+		if err := netlink.LinkSetUp(toIF); err != nil {
+			return err
+		}
+		if err := removeUplinkIPRoutes(fromIF); err != nil {
+			return err
+		}
+	}
+	toIndex := toIF.Attrs().Index
+	// Copy the uplink interface's IP to the veth interface.
+	if err := ConfigureLinkAddresses(toIndex, ips); err != nil {
+		return err
+	}
+	// Copy the uplink interface's routes to the veth interface.
+	if err := copyRoutes(toIF, routes); err != nil {
+		return err
+	}
+	return nil
+}
+
+// removeUplinkIPRoutes removes the IP and routes from the uplink interface.
+func removeUplinkIPRoutes(uplinkIF netlink.Link) error {
```

Comment:
ditto, this is in util package, it shouldn't even care whether this is an unlink, and the implementation has nothing to do with uplink.
And is it really worth to have `removeUplinkIPRoutes` call `removeRoutesOnLink`? I feel just naming this function as `removeAddrsOnLink` and calling `removeAddrsOnLink` and `removeRoutesOnLink` is more clear.